Add forward AD codegen support for single formula returning multiple outputs

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #75583

This PR:
- Updates `ForwardDerivative` to have tuples for the `var_types` and `var_names` fields (just like `Derivative`). Then dealt with the consequences.
- Disallow multiple-outputs when the return types includes a list type or the op is in-place

Codegen output:
```cpp
  c10::optional<::std::tuple<at::Tensor,at::Tensor,at::Tensor>> U_S_Vh_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_U_S_Vh) {
      auto A_t_raw = toNonOptFwGrad(A);
      auto A_tensor = toNonOptTensor(A);
      auto A_t = (A_t_raw.defined() || !A_tensor.defined())
        ? A_t_raw : at::_efficientzerotensor(A_tensor.sizes(), A_tensor.options());
      U_S_Vh_new_fw_grad_opt = linalg_svd_jvp(A_t, U, S, Vh, full_matrices);
  }
  if (U_S_Vh_new_fw_grad_opt.has_value() && std::get<0>(U_S_Vh_new_fw_grad_opt.value()).defined()) {
    U._set_fw_grad(std::get<0>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (U_S_Vh_new_fw_grad_opt.has_value() && std::get<1>(U_S_Vh_new_fw_grad_opt.value()).defined()) {
    S._set_fw_grad(std::get<1>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (U_S_Vh_new_fw_grad_opt.has_value() && std::get<2>(U_S_Vh_new_fw_grad_opt.value()).defined()) {
    Vh._set_fw_grad(std::get<2>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  if (grad_fn) {
    grad_fn->U_ = SavedVariable(U, true);
    grad_fn->S_ = SavedVariable(S, true);
    grad_fn->Vh_ = SavedVariable(Vh, true);
  }
  return std::make_tuple(std::move(U), std::move(S), std::move(Vh));
```

soulitzer committed:
Add forward AD codegen support for single formula returning multiple outputs

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add forward AD codegen support for single formula returning multiple outputs"


Codegen output:
```cpp
  c10::optional<::std::tuple<at::Tensor,at::Tensor,at::Tensor>> U_S_Vh_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_U_S_Vh) {
      auto A_t_raw = toNonOptFwGrad(A);
      auto A_tensor = toNonOptTensor(A);
      auto A_t = (A_t_raw.defined() || !A_tensor.defined())
        ? A_t_raw : at::_efficientzerotensor(A_tensor.sizes(), A_tensor.options());
      U_S_Vh_new_fw_grad_opt = linalg_svd_jvp(A_t, U, S, Vh, full_matrices);
  }
  if (U_S_Vh_new_fw_grad_opt.has_value()) {
    U._set_fw_grad(std::get<0>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (U_S_Vh_new_fw_grad_opt.has_value()) {
    S._set_fw_grad(std::get<1>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (U_S_Vh_new_fw_grad_opt.has_value()) {
    Vh._set_fw_grad(std::get<2>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  if (grad_fn) {
    grad_fn->U_ = SavedVariable(U, true);
    grad_fn->S_ = SavedVariable(S, true);
    grad_fn->Vh_ = SavedVariable(Vh, true);
  }
  return std::make_tuple(std::move(U), std::move(S), std::move(Vh));
}
```

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add forward AD codegen support for single formula returning multiple outputs"


Codegen output:
```cpp
  c10::optional<::std::tuple<at::Tensor,at::Tensor,at::Tensor>> U_S_Vh_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_U_S_Vh) {
      auto A_t_raw = toNonOptFwGrad(A);
      auto A_tensor = toNonOptTensor(A);
      auto A_t = (A_t_raw.defined() || !A_tensor.defined())
        ? A_t_raw : at::_efficientzerotensor(A_tensor.sizes(), A_tensor.options());
      U_S_Vh_new_fw_grad_opt = linalg_svd_jvp(A_t, U, S, Vh, full_matrices);
  }
  if (U_S_Vh_new_fw_grad_opt.has_value()) {
    U._set_fw_grad(std::get<0>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (U_S_Vh_new_fw_grad_opt.has_value()) {
    S._set_fw_grad(std::get<1>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (U_S_Vh_new_fw_grad_opt.has_value()) {
    Vh._set_fw_grad(std::get<2>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  if (grad_fn) {
    grad_fn->U_ = SavedVariable(U, true);
    grad_fn->S_ = SavedVariable(S, true);
    grad_fn->Vh_ = SavedVariable(Vh, true);
  }
  return std::make_tuple(std::move(U), std::move(S), std::move(Vh));
}
```

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add forward AD codegen support for single formula returning multiple outputs"


Codegen output:
```cpp
  c10::optional<::std::tuple<at::Tensor,at::Tensor,at::Tensor>> U_S_Vh_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_U_S_Vh) {
      auto A_t_raw = toNonOptFwGrad(A);
      auto A_tensor = toNonOptTensor(A);
      auto A_t = (A_t_raw.defined() || !A_tensor.defined())
        ? A_t_raw : at::_efficientzerotensor(A_tensor.sizes(), A_tensor.options());
      U_S_Vh_new_fw_grad_opt = linalg_svd_jvp(A_t, U, S, Vh, full_matrices);
  }
  if (U_S_Vh_new_fw_grad_opt.has_value()) {
    U._set_fw_grad(std::get<0>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (U_S_Vh_new_fw_grad_opt.has_value()) {
    S._set_fw_grad(std::get<1>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (U_S_Vh_new_fw_grad_opt.has_value()) {
    Vh._set_fw_grad(std::get<2>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  if (grad_fn) {
    grad_fn->U_ = SavedVariable(U, true);
    grad_fn->S_ = SavedVariable(S, true);
    grad_fn->Vh_ = SavedVariable(Vh, true);
  }
  return std::make_tuple(std::move(U), std::move(S), std::move(Vh));
}
```

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add forward AD codegen support for single formula returning multiple outputs"


This PR:
- Updates `ForwardDerivative` to have tuples for the `var_types` and `var_names` fields (just like `Derivative`). Then dealt with the consequences.
- Disallow multiple-outputs when the return types includes a list type or the op is in-place

Codegen output:
```cpp
  c10::optional<::std::tuple<at::Tensor,at::Tensor,at::Tensor>> U_S_Vh_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_U_S_Vh) {
      auto A_t_raw = toNonOptFwGrad(A);
      auto A_tensor = toNonOptTensor(A);
      auto A_t = (A_t_raw.defined() || !A_tensor.defined())
        ? A_t_raw : at::_efficientzerotensor(A_tensor.sizes(), A_tensor.options());
      U_S_Vh_new_fw_grad_opt = linalg_svd_jvp(A_t, U, S, Vh, full_matrices);
  }
  if (U_S_Vh_new_fw_grad_opt.has_value()) {
    U._set_fw_grad(std::get<0>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (U_S_Vh_new_fw_grad_opt.has_value()) {
    S._set_fw_grad(std::get<1>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (U_S_Vh_new_fw_grad_opt.has_value()) {
    Vh._set_fw_grad(std::get<2>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  if (grad_fn) {
    grad_fn->U_ = SavedVariable(U, true);
    grad_fn->S_ = SavedVariable(S, true);
    grad_fn->Vh_ = SavedVariable(Vh, true);
  }
  return std::make_tuple(std::move(U), std::move(S), std::move(Vh));
}
```

[ghstack-poisoned]

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -338,6 +338,12 @@
 """)
 
+FW_DERIVATIVE_SETTER_MULTI_OUTPUT = CodeTemplate("""\
+if (${all_res}_new_fw_grad_opt.has_value() && std::get<${idx}>(${all_res}_new_fw_grad_opt.value()).defined()) {
```
Is the code sample in the PR description up to date with this? I thing the defined check is missing from there?

soulitzer replied:
Updated!

albanD commented on tools/autograd/load_derivatives.py:
```diff
@@ -92,34 +92,40 @@ def create_derivative(f: NativeFunction, formula: str, var_names: Tuple[str, ...
         required_original_self_value=False,
         is_reusing_outplace_formula=False)
 
+    return fwd_deriv
```
Is there any reason to move this here?

soulitzer replied:
Nope, just wanted to print it at one point

albanD commented on tools/autograd/load_derivatives.py:
```diff
@@ -253,14 +260,12 @@ def repl(m: Any) -> str:
-    else:
-        return False
+    for name in names:
+        if name not in all_arg_names:
```
Is it possible to have an input and an output with the same name?

soulitzer replied:
Yes, the two ops where this happens are `symeig` and `eig`. Neither actually have a formula wrt that argument since it is a boolean. I just added an assert that checks that the name is not both input/output - both are deprecated though so no one should be trying to add forward AD formulas for those outputs.

albanD reviewed (commented):
Looks good!
Only nit.

soulitzer committed:
Update on "Add forward AD codegen support for single formula returning multiple outputs"


This PR:
- Updates `ForwardDerivative` to have tuples for the `var_types` and `var_names` fields (just like `Derivative`). Then dealt with the consequences.
- Disallow multiple-outputs when the return types includes a list type or the op is in-place

Codegen output:
```cpp
  c10::optional<::std::tuple<at::Tensor,at::Tensor,at::Tensor>> U_S_Vh_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_U_S_Vh) {
      auto A_t_raw = toNonOptFwGrad(A);
      auto A_tensor = toNonOptTensor(A);
      auto A_t = (A_t_raw.defined() || !A_tensor.defined())
        ? A_t_raw : at::_efficientzerotensor(A_tensor.sizes(), A_tensor.options());
      U_S_Vh_new_fw_grad_opt = linalg_svd_jvp(A_t, U, S, Vh, full_matrices);
  }
  if (U_S_Vh_new_fw_grad_opt.has_value() && std::get<0>(U_S_Vh_new_fw_grad_opt.value()).defined()) {
    U._set_fw_grad(std::get<0>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (U_S_Vh_new_fw_grad_opt.has_value() && std::get<1>(U_S_Vh_new_fw_grad_opt.value()).defined()) {
    S._set_fw_grad(std::get<1>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (U_S_Vh_new_fw_grad_opt.has_value() && std::get<2>(U_S_Vh_new_fw_grad_opt.value()).defined()) {
    Vh._set_fw_grad(std::get<2>(U_S_Vh_new_fw_grad_opt.value()), /* level */ 0, /* is_inplace_op */ false);
  }
  if (grad_fn) {
    grad_fn->U_ = SavedVariable(U, true);
    grad_fn->S_ = SavedVariable(S, true);
    grad_fn->Vh_ = SavedVariable(Vh, true);
  }
  return std::make_tuple(std::move(U), std::move(S), std::move(Vh));
```

[ghstack-poisoned]

albanD reviewed (approved):
SGTM!

soulitzer commented:
@pytorchbot merge this
